src/representation_theory/rep_hom.lean

```diff
@@ -0,0 +1,589 @@
+  {ρ : representation k G V} {ρ₂ : representation k G V₂}
+
+instance : rep_hom_class (ρ →ᵣ ρ₂) ρ ρ₂ :=
+{ coe := rep_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_add := rep_hom.map_add',
+  map_smulₛₗ := rep_hom.map_smul',
+  map_smulG := rep_hom.map_smulG' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
+directly.
+-/
+instance : has_coe_to_fun (ρ →ᵣ ρ₂) (λ _, V → V₂) := ⟨λ f, f⟩
+-- should there be coe to linear_map?
+@[simp] lemma to_fun_eq_coe {f : ρ →ᵣ ρ₂} : f.to_fun = (f : V → V₂) := rfl
+
+@[simp] lemma coe_eq_to_linear_map_coe {f : ρ →ᵣ ρ₂} :
+  (f.to_linear_map : V → V₂) = (f : V → V₂) := rfl
+
+@[ext] theorem ext {f g : ρ →ᵣ ρ₂} (h : ∀ x, f x = g x) : f = g := fun_like.ext f g h
+
+/-- Copy of a `rep_hom` with a new `to_fun` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def copy (f : ρ →ᵣ ρ₂) (f' : V → V₂) (h : f' = ⇑f) : ρ →ᵣ ρ₂ :=
+{ to_fun := f',
+  map_add' := h.symm ▸ f.map_add',
+  map_smul' := h.symm ▸ f.map_smul',
+  map_smulG' := h.symm ▸ f.map_smulG' }
+
+@[simp] lemma coe_mk (f : V → V₂) (h₁ h₂ h₃) :
+  ((rep_hom.mk f h₁ h₂ h₃ : ρ →ᵣ ρ₂) : V → V₂) = f := rfl
+
+/-- Identity map as a `rep_hom` -/
+def id : ρ →ᵣ ρ :=
+{ to_fun := id,
+  map_smulG' := by simp,
+  ..linear_map.id }
+
+lemma id_apply (x : V) :
+  @id k G V _ _ _ _ ρ x = x := rfl
+
+@[simp, norm_cast] lemma id_coe : ((rep_hom.id : ρ →ᵣ ρ) : V → V) = _root_.id := rfl
+
+section
+theorem rep_hom_is_rep_hom (f : ρ →ᵣ ρ₂) : is_rep_hom ρ ρ₂ f :=
+⟨f.map_add', f.map_smul', f.map_smulG'⟩
+
+variables {f f' : ρ →ᵣ ρ₂}
+
+theorem coe_injective : @injective (ρ →ᵣ ρ₂) (V → V₂) coe_fn :=
+fun_like.coe_injective
+
+protected lemma congr_arg {x x' : V} : x = x' → f x = f x' :=
+fun_like.congr_arg f
+
+protected lemma congr_fun (h : f = f') (x : V) : f x = f' x :=
+fun_like.congr_fun h x
+
+theorem ext_iff : f = f' ↔ ∀ x, f x = f' x :=
+fun_like.ext_iff
+
+@[simp] lemma mk_coe (f : ρ →ᵣ ρ₂) (h₁ h₂ h₃) :
+  (rep_hom.mk f h₁ h₂ h₃ : ρ →ᵣ ρ₂) = f := ext $ λ _, rfl
+
+variables (f f')
+
+protected lemma map_add (x y : V) : f (x + y) = f x + f y := map_add f x y
+protected lemma map_zero : f 0 = 0 := map_zero f
+protected lemma map_smul (c : k) (x : V) : f (c • x) = c • f x := map_smul f c x
+protected lemma map_smulG (g : G) (x : V) : f (ρ g x) = ρ₂ g (f x) := map_smulG f g x
+
+@[simp] lemma map_eq_zero_iff (h : function.injective f) {x : V} : f x = 0 ↔ x = 0 :=
+⟨λ w, by { apply h, simp [w], }, λ w, by { subst w, simp, }⟩
+
+section pointwise
+open_locale pointwise
+
+@[simp] lemma image_smul_set (c : k) (s : set V) :
+  f '' (c • s) = c • f '' s :=
+begin
+  apply set.subset.antisymm,
+  { rintros x ⟨y, ⟨z, zs, rfl⟩, rfl⟩,
+    exact ⟨f z, set.mem_image_of_mem _ zs, (f.map_smul _ _).symm ⟩ },
+  { rintros x ⟨y, ⟨z, hz, rfl⟩, rfl⟩,
+    exact (set.mem_image _ _ _).2 ⟨c • z, set.smul_mem_smul_set hz, f.map_smul _ _⟩ }
+end
```

Comment:
Does `f.to_linear_map.image_smul_set` works? Same for preimage below.

Reply:
Yep, it works. Same for preimage